types/node.lua

```diff
@@ -479,157 +568,277 @@ function nodetypes.alternative:deltas ()
 -- entangled and point towards a more general design issue.
 -- N.B. this vkern node is only used in the linespacing package so far.
-nodetypes.vkern.discardable = false
+vkern.discardable = false
 
-function nodetypes.vkern:__tostring ()
+function vkern:__tostring ()
    return "VK<" .. tostring(self.height) .. ">"
 end
 
-nodetypes.penalty = pl.class(nodetypes.box)
-nodetypes.penalty.type = "penalty"
-nodetypes.penalty.discardable = true
-nodetypes.penalty.penalty = 0
-
-function nodetypes.penalty:_init (spec)
+--- A penalty node has a value which is used by the line breaking algorithm (in horizontal mode)
+-- or the page breaking algorithm (in vertical mode), to determine where to break.
+-- The value is expected to be a number between -10000 and 10000.
+-- The higher the value, the less desirable it is to break at that point.
+-- The exterme values (-10000 and 10000) are used to indicate that the break is forbidden or mandatory,
+-- i.e. in certain way represent an infinite penalty.
+--
+-- Derived from `box`.
+--
+-- Property is_penalty is true.
+--
+-- @type penalty
+
+local penalty = pl.class(box)
+penalty.type = "penalty"
+penalty.discardable = true
+penalty.penalty = 0
+
+--- Constructor
+-- @tparam table spec A table with the properties of the penalty.
+function penalty:_init (spec)
    self:super(spec)
    if type(spec) ~= "table" then
       self.penalty = SU.cast("number", spec)
    end
 end
 
-function nodetypes.penalty:__tostring ()
+function penalty:__tostring ()
    return "P(" .. tostring(self.penalty) .. ")"
 end
 
-function nodetypes.penalty.outputYourself (_) end
+--- Output routine for a penalty.
+-- If found in the ouptut, penalties have no representation, so this method does nothing.
+-- (Overriding it on some penalties may be useful for debugging purposes.)
+--
+function penalty.outputYourself (_, _, _) end
 
-function nodetypes.penalty.toText (_)
+--- Returns a text representation of the penalty node.
+-- This is used for debugging purposes, returning '(!)' for a penalty.
+--
+-- @treturn string A string representation of the penalty node ('(!)').
+function penalty.toText (_)
    return "(!)"
 end
 
-function nodetypes.penalty:unbox ()
+--- Unbox a penalty.
+-- This method exists consistency with vbox-derived classes, for a penalty used in vertical mode.
+--
+-- @treturn table A table with the penalty node.
+function penalty:unbox ()
    return { self }
 end
 
-nodetypes.vbox = pl.class(nodetypes.box)
-nodetypes.vbox.type = "vbox"
-nodetypes.vbox.nodes = {}
-nodetypes.vbox._default_length = "height"
+--- A vbox is a box node used in vertical mode.
+--
+-- Derived from `box`.
+--
+-- Properties is_vbox and is_box are true.
+--
+-- @type vbox
 
-function nodetypes.vbox:_init (spec)
+local vbox = pl.class(box)
+vbox.type = "vbox"
+vbox.nodes = {}
+vbox._default_length = "height"
+
+--- Constructor
+--
+-- @tparam table spec A table with the properties of the vbox.
+function vbox:_init (spec)
```

Comment:
I remember having suffered here. As with all of these classes, one can pass almost anything and the box class makes it a field. So if called with `SILE.types.node.vbox({ nodes = ... }` then the nodes will be set after the parent class super call. Ok... and it's indeed used this way with lines in the typesetter.
  - So why does pagebuilder.collateVboxes uses an empty constructor and an explicit append()?
  - ... Oh wait, the depth/height logic then differs (here, the max, but with append it's  "the box height is the height is the total height of the content, minus the depth of the last box."

If the discrepancy is legit (?) then some explanation is needed, 'cause again one can't make any sense of it currently.